File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.java
import java.util.*;

class Solution {
    
    private static final int[] dr = {-1, 0, 0, 1};
    private static final int[] dc = {0, -1, 1, 0};
    
    public int solution(int[][] game_board, int[][] table) {
        
        int[][][] boardCoords = findCoords(game_board, 0);
        int[][][] tableCoords = findCoords(table, 1);

		int[][][] boardMatrices = new int[boardCoords.length][][];
        for (int i = 0; i < boardMatrices.length; i++) {
            boardMatrices[i] = makeMatrix(boardCoords[i]);
        }
        
        int[][][] tableMatrices = new int[tableCoords.length][][];
        for (int i = 0; i < tableMatrices.length; i++) {
            tableMatrices[i] = makeMatrix(tableCoords[i]);
        }
        
        int answer = 0;

        for (int[][] tableMatrix : tableMatrices) {
            rotateLoop: for (int i = 0; i < 4; i++) {
                tableMatrix = rotate(tableMatrix);

                boardLoop: for (int j = 0; j < boardMatrices.length; j++) {
                    if (boardMatrices[j] == null) continue;

                    if (tableMatrix.length != boardMatrices[j].length || tableMatrix[0].length != boardMatrices[j][0].length)
                        continue;

                    int count = 0;
                    for (int r = 0; r < tableMatrix.length; r++) {
                        for (int c = 0; c < tableMatrix[0].length; c++) {
                            if (tableMatrix[r][c] != boardMatrices[j][r][c]) continue boardLoop;
                            if (tableMatrix[r][c] == 1) count++;
                        }
                    }

                    answer += count;
                    boardMatrices[j] = null;
                    break rotateLoop;
                }
            }
        }
        
       	return answer;
    }
    
    private int[][][] findCoords(int[][] board, int target) {
        int N = board.length;
        
        List<List<int[]>> coords = new ArrayList<>();
        boolean[][] visited = new boolean[N][N];
        Deque<int[]> deq = new ArrayDeque<>();
        
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                
                if (visited[i][j] || board[i][j] != target) continue;
                
                List<int[]> crntCoords = new ArrayList<>();
                deq.add(new int[]{i, j});
                
                while (!deq.isEmpty()) {
                    
                    int[] crnt = deq.poll();
                    
                    int r = crnt[0];
                    int c = crnt[1];
                    
                    visited[r][c] = true;
                    crntCoords.add(new int[]{r, c});
                    
                    for (int k = 0; k < 4; k++) {
                        int nr = r + dr[k];
                        int nc = c + dc[k];
                        if (nr < 0 || nc < 0 || nr >= N || nc >= N) continue;
                        if (visited[nr][nc] || board[nr][nc] != target) continue;
                        
						deq.add(new int[]{nr, nc});
                    }
                }
                
                if (!crntCoords.isEmpty()) coords.add(crntCoords);
            }
        }
        
        return coords.stream()
            .map(list -> list.toArray(new int[0][]))
            .toArray(int[][][]::new);
    }
    
    private int[][] makeMatrix(int[][] coords) {
        int rowMin = Integer.MAX_VALUE;
        int colMin = Integer.MAX_VALUE;
        for (int[] coord: coords) {
            rowMin = Math.min(coord[0], rowMin);
            colMin = Math.min(coord[1], colMin);
        }
        
        for (int[] coord: coords) {
            coord[0] -= rowMin;
            coord[1] -= colMin;
        }
        
        int rowMax = Integer.MIN_VALUE;
        int colMax = Integer.MIN_VALUE;
        for (int[] coord: coords) {
            rowMax = Math.max(coord[0], rowMax);
            colMax = Math.max(coord[1], colMax);
        }

        int[][] arr = new int[rowMax + 1][colMax + 1];
        for (int[] coord: coords) {
            int r = coord[0];
            int c = coord[1];
            arr[r][c] = 1;
        }
        
        return arr;
    }
    
    private int[][] rotate(int[][] original) {
        
        int N = original.length;
        int M = original[0].length;
        
        int[][] rotated = new int[M][N];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) rotated[j][i] = original[i][j];
        }
        
        for (int i = 0; i < M / 2; i++) {
            int[] temp = rotated[i];
            rotated[i] = rotated[M - 1 - i];
            rotated[M - 1 - i] = temp;
        }
        
        return rotated;
    }
}

File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.py
from collections import deque

dr = [-1, 0, 0, 1]
dc = [0, -1, 1, 0]


def find_blocks(board, target):
    n = len(board)
    visited = [[False] * n for _ in range(n)]
    result = []

    for r in range(n):
        for c in range(n):
            if board[r][c] != target or visited[r][c]:
                continue

            queue = deque([(r, c)])
            visited[r][c] = True
            block = []

            while queue:
                current_r, current_c = queue.popleft()
                block.append((current_r, current_c))

                for direction in range(4):
                    nr = current_r + dr[direction]
                    nc = current_c + dc[direction]

                    if nr < 0 or nc < 0 or nr >= n or nc >= n:
                        continue

                    if visited[nr][nc] or board[nr][nc] != target:
                        continue

                    visited[nr][nc] = True
                    queue.append((nr, nc))

            result.append(normalize_block(block))

    return result


def normalize_block(block):
    min_r = min(r for r, c in block)
    min_c = min(c for r, c in block)

    return sorted(
        (r - min_r, c - min_c)
        for r, c in block
    )


def rotate(block):
    # (r, c) -> (c, -r)로 90도 회전
    rotated = [(c, -r) for r, c in block]

    # 회전하면 좌표의 기준점과 순서가 달라지므로 다시 정규화
    return normalize_block(rotated)


def solution(game_board, table):
    empties = find_blocks(game_board, 0)
    blocks = find_blocks(table, 1)

    used = [False] * len(blocks)
    answer = 0

    for empty in empties:
        for index, block in enumerate(blocks):
            if used[index]:
                continue

            # 칸 수가 다르면 모양이 같을 수 없음
            if len(empty) != len(block):
                continue

            rotated_block = block
            matched = False

            # 0도, 90도, 180도, 270도
            for _ in range(4):
                if empty == rotated_block:
                    used[index] = True
                    answer += len(block)
                    matched = True
                    break

                rotated_block = rotate(rotated_block)

            if matched:
                break

    return answer